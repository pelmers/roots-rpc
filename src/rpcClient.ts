import * as t from "io-ts";
import { isRight } from "fp-ts/lib/Either";

import {
  RpcTransport,
  AnyJson,
  AsyncFN,
  Disposable,
  ErrorPayload,
  FNDecl,
  ResultPayload,
  Observable,
  disposify,
  Message,
} from "./rpcTypes";

class Resolver<T> {
  public resolve: (t: T) => void;
  public reject: (err: unknown) => void;
  public promise: Promise<T>;
  constructor() {
    this.promise = new Promise<T>((resolve, reject) => {
      this.resolve = resolve;
      this.reject = reject;
    });
  }
}

type Subscription<T> = {
  type: t.Type<T>;
  name: string;
  onValue: (value: T) => void;
  onError?: (err: Error) => void;
  onComplete?: () => void;
};

export class RpcClient {
  private id: number = 0;
  private disposable: Disposable;
  private pendingCalls: Map<number, Resolver<AnyJson>> = new Map();
  private subscriptions: Map<number, Subscription<AnyJson>> = new Map();

  constructor(private transport: RpcTransport) {
    this.disposable = transport.onMessage((msg) => {
      const { id } = msg;
      if (this.pendingCalls.has(id)) {
        this.handleCall(msg);
      } else if (this.subscriptions.has(id)) {
        this.handleSub(msg);
      }
    });
  }

  private handleCall(msg: Message) {
    const { id } = msg;
    const { resolve, reject } = this.pendingCalls.get(id)!;
    if ("error" in msg) {
      reject(msg.error);
    } else if ("result" in msg) {
      resolve(msg.result);
    }
  }

  private handleSub(msg: Message) {
    const { id } = msg;
    const sub = this.subscriptions.get(id)!;
    if ("error" in msg) {
      sub.onError && sub.onError(new Error(msg.error));
    } else if ("result" in msg) {
      try {
        this.checkReturn(id, sub.name, sub.type, msg.result);
      } catch (e) {
        sub.onError && sub.onError(e);
        return;
      }
      sub.onValue(msg.result);
    } else if ("dispose" in msg) {
      sub.onComplete && sub.onComplete();
      this.subscriptions.delete(id);
    }
  }

  private checkInput<I>(name: string, i: t.Type<I>, value?: AnyJson) {
    if (value === undefined && !i.is(null)) {
      throw new Error(
        `No argument passed for non-null input of RPC function ${name}`
      );
    }
  }

  private checkReturn<O>(
    id: number,
    name: string,
    o: t.Type<O>,
    value: AnyJson
  ) {
    const validation = o.decode(value);
    if (!isRight(validation)) {
      throw new Error(
        `Failed to validate return type of call id ${id}, method ${name}`
      );
    }
  }

  connect<I extends AnyJson, O extends AnyJson>(
    typeLoader: () => FNDecl<I, O>
  ): AsyncFN<I, O> {
    return async (argValue?: I) => {
      const { name } = typeLoader;
      const { i, o } = typeLoader();
      const id = this.id++;
      this.checkInput(name, i, argValue);
      const arg = argValue === undefined ? null : argValue;
      const pendingCall = new Resolver<O>();
      this.pendingCalls.set(id, pendingCall);
      this.transport.send({ id, method: name, arg, subscribe: false });
      let result;
      try {
        result = await pendingCall.promise;
      } finally {
        this.pendingCalls.delete(id);
      }
      this.checkReturn(id, name, o, result);
      return result;
    };
  }

  connectObservable<I extends AnyJson, O extends AnyJson>(
    typeLoader: () => FNDecl<I, O>
  ): (arg?: I) => Observable<O> {
    return (argValue?: I) => {
      const { name } = typeLoader;
      const { i, o } = typeLoader();
      const id = this.id++;
      this.checkInput(name, i, argValue);
      return {
        subscribe: (onValue, onError?, onComplete?) => {
          const arg = argValue === undefined ? null : argValue;
          this.subscriptions.set(id, {
            name,
            type: o,
            onValue,
            onError,
            onComplete,
          });
          this.transport.send({ id, subscribe: true, method: name, arg });
          return disposify(() => {
            this.transport.send({ id, dispose: true });
            onComplete && onComplete();
            this.subscriptions.delete(id);
          });
        },
      };
    };
  }

  dispose() {
    this.disposable.dispose();
    this.pendingCalls.clear();
    this.subscriptions.clear();
  }
}
