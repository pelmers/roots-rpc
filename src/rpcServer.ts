import { isRight } from "fp-ts/lib/Either";
import {
  RpcTransport,
  AnyJson,
  FNDecl,
  Disposable,
  CallPayload,
  disposify,
  Observable,
} from "./rpcTypes";

type Record =
  | {
      fn: (payload: AnyJson) => Promise<AnyJson>;
      type: FNDecl<AnyJson, AnyJson>;
    }
  | {
      start: (payload: AnyJson) => Observable<AnyJson>;
      type: FNDecl<AnyJson, AnyJson>;
    };

export class RpcServer {
  private registry: {
    [key: string]: Record;
  } = {};
  private subscriptions: Map<number, Disposable> = new Map();
  private disposables: Disposable[] = [];

  constructor(private transport: RpcTransport) {
    this.disposables.push(
      transport.onMessage((msg) => {
        if ("method" in msg) {
          this.handleCallMessage(msg);
        } else if ("dispose" in msg) {
          const subscription = this.subscriptions.get(msg.id);
          subscription?.dispose();
          this.subscriptions.delete(msg.id);
        }
      })
    );
  }

  private handleCallMessage(msg: CallPayload): void {
    const { id, method, arg, subscribe } = msg;
    const record = this.registry[method];
    if (record == null) {
      this.transport.send({
        id,
        error: `Handler implementation not registered for method ${method} (id: ${id})`,
      });
      return;
    }
    const validation = record.type.i.decode(arg);
    if (!isRight(validation)) {
      this.transport.send({
        id,
        error: `Input validation failed for call id ${id} on method ${method}`,
      });
      return;
    }

    if ("fn" in record) {
      if (subscribe) {
        this.transport.send({
          id,
          error: `Can not subscribe to regular function method ${method}`,
        });
        return;
      }
      record
        .fn(arg)
        .then((result) => this.transport.send({ id, result }))
        .catch((err) => this.transport.send({ id, error: err.toString() }));
    } else if ("start" in record) {
      if (!subscribe) {
        this.transport.send({
          id,
          error: `Must subscribe to observable method ${method}`,
        });
        return;
      }
      this.subscriptions.set(
        id,
        record.start(arg).subscribe(
          (v) => this.transport.send({ id, result: v }),
          (e) => this.transport.send({ id, error: e.toString() }),
          () => {
            this.transport.send({ id, dispose: true });
            this.subscriptions.delete(id);
          }
        )
      );
    }
  }

  private enforceNameAvailable(name: string) {
    if (this.registry.hasOwnProperty(name)) {
      throw new Error(`registry already has handler registered for ${name}`);
    }
  }

  register<I extends AnyJson, O extends AnyJson>(
    typeLoader: () => FNDecl<I, O>,
    impl: (i: I) => Promise<O>
  ): Disposable {
    this.enforceNameAvailable(typeLoader.name);
    this.registry[typeLoader.name] = { type: typeLoader(), fn: impl };
    return disposify(() => delete this.registry[typeLoader.name]);
  }

  registerObservable<I extends AnyJson, O extends AnyJson>(
    typeLoader: () => FNDecl<I, O>,
    impl: (i: I) => Observable<O>
  ): Disposable {
    this.enforceNameAvailable(typeLoader.name);
    this.registry[typeLoader.name] = { type: typeLoader(), start: impl };
    return disposify(() => delete this.registry[typeLoader.name]);
  }

  dispose() {
    this.disposables.forEach((d) => d.dispose());
    for (const [id, sub] of this.subscriptions) {
      this.transport.send({ id, dispose: true });
      sub.dispose();
    }
  }
}
