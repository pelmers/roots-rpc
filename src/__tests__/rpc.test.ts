import EventEmitter from "events";

import * as t from "io-ts";

import { RpcServer } from "../rpcServer";
import {
  Message,
  AsyncFN,
  Disposable,
  RpcTransport,
  Observable,
  disposify,
} from "../rpcTypes";
import { RpcClient } from "../rpcClient";

class TestTransport implements RpcTransport {
  constructor(public sender: EventEmitter, public receiver: EventEmitter) {}
  send(msg: Message) {
    this.sender.emit("msg", msg);
  }
  onMessage(cb: (msg: Message) => void) {
    this.receiver.on("msg", cb);
    return { dispose: () => this.receiver.off("msg", cb) };
  }
}

class TestObservable implements Observable<string> {
  constructor(private value: string, private time: number) {}
  subscribe(
    onValue: (value: string) => void,
    _onError?: (err: Error) => void,
    onComplete?: () => void
  ): Disposable {
    let disposed = false;
    const emitValues = async () => {
      for (const char of this.value) {
        if (disposed) {
          return;
        }
        onValue(char);
        await new Promise((resolve) => setTimeout(resolve, this.time));
      }
      onComplete && onComplete();
    };
    emitValues();
    return disposify(() => (disposed = true));
  }
}

const TestIn = t.type({ a: t.number, b: t.number });
const TestOut = t.type({ x: t.string, y: t.number });

function TestRpcMethod() {
  return { i: TestIn, o: TestOut };
}

function TestNullMethod() {
  return { i: t.null, o: t.null };
}

const TestObservableIn = t.type({ val: t.string, time: t.number });
function TestObservableMethod() {
  return { i: TestObservableIn, o: t.string };
}

describe("rpc", () => {
  let serverEmitter: EventEmitter;
  let clientEmitter: EventEmitter;
  let serverTransport: TestTransport;
  let clientTransport: TestTransport;
  let server: RpcServer;
  let client: RpcClient;
  let rpcHandler: Disposable;
  let rpcMethod: AsyncFN<t.TypeOf<typeof TestIn>, t.TypeOf<typeof TestOut>>;
  let rpcObservableSource: Disposable;
  let rpcObservable: (
    arg?: t.TypeOf<typeof TestObservableIn>
  ) => Observable<string>;

  beforeEach(() => {
    serverEmitter = new EventEmitter();
    clientEmitter = new EventEmitter();
    serverTransport = new TestTransport(clientEmitter, serverEmitter);
    clientTransport = new TestTransport(serverEmitter, clientEmitter);
    server = new RpcServer(serverTransport);
    client = new RpcClient(clientTransport);
    rpcHandler = server.register(TestRpcMethod, async ({ a, b }) => ({
      x: "abcd".slice(a),
      y: a * b,
    }));
    rpcMethod = client.connect(TestRpcMethod);

    rpcObservableSource = server.registerObservable(
      TestObservableMethod,
      (i) => new TestObservable(i.val, i.time)
    );
    rpcObservable = client.connectObservable(TestObservableMethod);
  });

  it("disposes the listeners in transport", () => {
    expect(serverEmitter.listenerCount("msg")).toEqual(1);
    expect(clientEmitter.listenerCount("msg")).toEqual(1);

    server.dispose();
    client.dispose();
    expect(serverEmitter.listenerCount("msg")).toEqual(0);
    expect(clientEmitter.listenerCount("msg")).toEqual(0);
  });

  it("gets correct result", async () => {
    const { x, y } = await rpcMethod({ a: 10, b: 5 });
    expect(x).toEqual("");
    expect(y).toEqual(50);
  });

  it("checks input type", async () => {
    await expect(
      // @ts-ignore ignore for testing
      rpcMethod({ a: "not a number", b: null })
    ).rejects.toMatchSnapshot();
  });

  it("requires input value", async () => {
    await expect(rpcMethod()).rejects.toMatchSnapshot();
  });

  it("checks return type", async () => {
    rpcHandler.dispose();
    // @ts-ignore ignore for testing
    server.register(TestRpcMethod, async () => ({
      x: 100,
      y: "not a number!",
    }));
    await expect(rpcMethod({ a: 10, b: 5 })).rejects.toMatchSnapshot();
  });

  it("allows empty arg for null", async () => {
    // @ts-ignore ignore for testing
    server.register(TestNullMethod, async () => null);
    const method = client.connect(TestNullMethod);
    const res = await method();
    expect(res).toBeNull();
  });

  it("runs an observable correctly", async () => {
    let errored = false;
    const str = "xyz123";
    const values: string[] = [];
    await new Promise<void>((resolve) => {
      rpcObservable({ val: str, time: 5 }).subscribe(
        (v) => values.push(v),
        () => (errored = true),
        resolve
      );
    });
    expect(values.join("")).toEqual(str);
    expect(errored).toBeFalsy();
  });

  it("lets client dispose observable", async () => {
    let errored = false;
    let completed = false;
    const str = "xyz123";
    const values: string[] = [];
    await new Promise<void>((resolve) => {
      const disposable = rpcObservable({ val: str, time: 30 }).subscribe(
        (v) => {
          values.push(v);
          if (values.length === str.length / 2) {
            disposable.dispose();
            resolve();
          }
        },
        () => (errored = true),
        () => (completed = true)
      );
    });
    expect(values.join("")).toEqual(str.slice(0, str.length / 2));
    expect(errored).toBeFalsy();
    expect(completed).toBeTruthy();
  });

  it("sends error of observable", async () => {
    let errored: Error | undefined;
    rpcObservableSource.dispose();
    rpcObservableSource = server.registerObservable(
      TestObservableMethod,
      (_) => ({
        subscribe(_, onError, onComplete) {
          onError!(new Error("Test error message!!"));
          onComplete!();
          return disposify(() => {});
        },
      })
    );
    await new Promise<void>((resolve) => {
      rpcObservable({ val: "", time: 0 }).subscribe(
        (_) => {},
        (e) => (errored = e),
        resolve
      );
    });
    expect(errored).toMatchSnapshot();
  });
});
