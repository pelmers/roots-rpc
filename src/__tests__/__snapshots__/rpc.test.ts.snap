// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rpc checks input type 1`] = `"Input validation failed for call id 0 on method TestRpcMethod"`;

exports[`rpc checks return type 1`] = `[Error: Failed to validate return type of call id 0, method TestRpcMethod]`;

exports[`rpc requires input value 1`] = `[Error: No argument passed for non-null input of RPC function TestRpcMethod]`;

exports[`rpc sends error of observable 1`] = `[Error: Error: Test error message!!]`;
